### PR TITLE
feat: add service user crd

### PR DESCRIPTION
**Pull request by kjuulh**
This is an addition to support the use-case where we may want a user spanning a database host. The user should be automatically created, with a given password and associated roles.

The user is granted basic settings, such as connect, password etc, such that it can operate as a service user, and access the database itself, with no permissions.

This feature is proposed because we've got a need for easily creating and managing roles across many databases with the same settings and behavior

To see the spec: https://github.com/lunarway/postgresql-controller/blob/834e937a26a097a5f5a4e46bd11a3b2c2ca9b707/api/v1alpha1/postgresqlserviceuser_types.go

We expect that this CRD would be used in this way:

1. A password is created manually (optionally through a secret manager of some kind)
2. The CRD is created pr database
3. The postgresql-controller will reconcile the users against the specified database, it will create the user if not exist, assign the password, grant the roles.
4. Profit.

We'd love some feedback on the spec itself, the workflow, do you see any problems in the way we want to reconcile the user etc.

### CODE REVIEW

File: api/v1alpha1/postgresqlserviceuser_types.go [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+type PostgreSQLServiceUserSpec struct {
+	// Name of the service user
+	Name string `json:"name"`
+
+	// Host to connect to
+	Host ResourceVar `json:"host"`
+
+	// Password used with the Host and Name used to connect to the database
+	// +optional
+	Password *ResourceVar `json:"password"`
```
Comment (mahlunar) [inline api/v1alpha1/postgresqlserviceuser_types.go:36]: Do we want this to be optional? I believe we always want to supply a password .

File: api/v1alpha1/postgresqlserviceuser_types.go [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   PostgreSQLServiceUserSpec   `json:"spec,omitempty"`
+	Status PostgreSQLServiceUserStatus `json:"status,omitempty"`
+}
+
+//+kubebuilder:object:root=true
+
+// PostgreSQLServiceUserList contains a list of PostgreSQLServiceUser
+type PostgreSQLServiceUserList struct {
```
Comment (mahlunar) [inline api/v1alpha1/postgresqlserviceuser_types.go:69]: This is not referenced in any CRD?

Comment (kjuulh) [reply api/v1alpha1/postgresqlserviceuser_types.go:69]: No idea, it was created via. the scaffold =D

**Comment by tmablunar** [comment]
It would be more inline/ready for the secrets-management setup if the CRD defined a `ServiceRole`. postgres-controller could then create the role (like with `management_role`) and Vault could create a User and grant this role to it.

Not really related to this PR but...
If you then expanded go-postgres with a `WithDatabase` and used that instead of `WithConnection` you believe you could more or less use the existing rotateSecrets functionality. If we need to change a few things in the secrets-management-controller we could help with that.

File: api/v1alpha1/postgresqlserviceuser_types.go (hunk 1 above)
Comment (kjuulh) [reply api/v1alpha1/postgresqlserviceuser_types.go:36]: We'll start by keeping it optional, as that will allow us to migrate to secret management if we choose to go that route.

File: api/v1alpha1/postgresqlserviceuser_types.go (hunk 2 above)
Comment (kjuulh) [reply api/v1alpha1/postgresqlserviceuser_types.go:99]: Nvm. It is registered down below in the schema builder. I see the same is done for the other files =D

**Review by mahlunar — APPROVED** [review]

File: api/v1alpha1/postgresqlserviceuser_types.go [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+	PhaseUpdated metav1.Time `json:"phaseUpdated"`
+
+	// Phase which state is the given resource currently in
+	Phase PostgreSQLServiceUserPhase `json:"phase"`
+
+	// Host, which host was reconciled
+	Host string `json:"host,omitempty"`
+
+	// Name, which service user was reconciled
+	Name string `json:"name,omitempty"`
```
Comment (JacobValdemar) [inline api/v1alpha1/postgresqlserviceuser_types.go:78]: The `Name` here could be changed based on your answer to the question above.

File: api/v1alpha1/postgresqlserviceuser_types.go [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+
+	// Error if present, how did the reconciliation loop fail
+	Error string `json:"error,omitempty"`
+}
```
Comment (JacobValdemar) [inline api/v1alpha1/postgresqlserviceuser_types.go:82]: The pattern of using phase is deprecated. You should use the `conditions` status property instead. See it used in pods, deployments, etc.:
```sh
k get po POD_NAME -oyaml # scroll to the bottom
```

Example implementation: https://github.com/bitnami-labs/sealed-secrets/blob/main/pkg/apis/sealedsecrets/v1alpha1/types.go#L100-L128

**Relevant** documentation (actually read this): https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#typical-status-properties

File: api/v1alpha1/postgresqlserviceuser_types.go [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+//+kubebuilder:object:root=true
+//+kubebuilder:subresource:status
+
+// PostgreSQLServiceUser is the Schema for the postgresqlserviceusers API
+type PostgreSQLServiceUser struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   PostgreSQLServiceUserSpec   `json:"spec,omitempty"`
+	Status PostgreSQLServiceUserStatus `json:"status,omitempty"`
```
Comment (JacobValdemar) [inline api/v1alpha1/postgresqlserviceuser_types.go:93]: If this is optional, do this:
```suggestion
	// +optional    
	Status PostgreSQLServiceUserStatus `json:"status,omitempty"`
```

More information: https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#optional-vs-required

File: api/v1alpha1/postgresqlserviceuser_types.go (hunk 3 above)
Comment (JacobValdemar) [inline api/v1alpha1/postgresqlserviceuser_types.go:78]: You want to remove `omitempty` if these fields are required. More information: https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#optional-vs-required

**Review by JacobValdemar — CHANGES_REQUESTED** [review]
Generally, make sure to not use `omitempty` on required fields.

Generally, check out the documentation on [API conventions](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md).

**Comment by hoeg** [comment]
> @tmablunar Good comments. I think we're gonna go the route where we define a service role, that may or may not have a password. This will allow us to piggy back on secret management once we're ready for that, and if it makes sense client side. Client side is where I see the issues. But we'll dig into that another time.

Would you be able to expand a bit on why it might not be possible to use it client side? Perhaps the clients are not running in kubernetes? The cost of changing this later might be greater than start using it from the start. I was just curious as to why using static credentials is not possible :)

**Comment by kjuulh** [comment]
> @hoeg Would you be able to expand a bit on why it might not be possible to use it client side? Perhaps the clients are not running in kubernetes? The cost of changing this later might be greater than start using it from the start. I was just curious as to why using static credentials is not possible :)

It isn't as much as it isn't possible to use secret management. But whether it is ergonomic, and the way this specific application we're creating is supposed to get credentials, which may or may not be suited for secret management (which is why we're deferring the decision until we've got a better idea of how we want it to access the databases and credentials). 

We also have to think about whether secret management should maintain static database credentials, rotate them, whether they should be created on behalf of the client application, or be created as part of our service role.

**Comment by tmablunar** [comment]
> > @hoeg Would you be able to expand a bit on why it might not be possible to use it client side? Perhaps the clients are not running in kubernetes? The cost of changing this later might be greater than start using it from the start. I was just curious as to why using static credentials is not possible :)
> 
> It isn't as much as it isn't possible to use secret management. But whether it is ergonomic, and the way this specific application we're creating is supposed to get credentials, which may or may not be suited for secret management (which is why we're deferring the decision until we've got a better idea of how we want it to access the databases and credentials).
> 
> We also have to think about whether secret management should maintain static database credentials, rotate them, whether they should be created on behalf of the client application, or be created as part of our service role.

I understand why you might want to defer the decision but IF you want to go with secrets-management I believe we would needs these changes:
* go-postgres needs an option to set the database without disabling dynamic secrets
* secrets-management-controller needs to expose an option to ask for a specific role. Doing this would also enable support for shared databases.

Both things, I believe, are quite manageable

File: api/v1alpha1/postgresqlserviceuser_types.go (hunk 5 above)
Comment (kjuulh) [reply api/v1alpha1/postgresqlserviceuser_types.go:93]: The kubebuilder scaffolder really should get updated, it is actually the one that wrote this

```
type PostgreSQLServiceUserSpec struct {
	// INSERT ADDITIONAL SPEC FIELDS - desired state of cluster
	// Important: Run "make" to regenerate code after modifying this file

	// Foo is an example field of PostgreSQLServiceUser. Edit postgresqlserviceuser_types.go to remove/update
	Foo string `json:"foo,omitempty"`
}

// PostgreSQLServiceUserStatus defines the observed state of PostgreSQLServiceUser
type PostgreSQLServiceUserStatus struct {
	// INSERT ADDITIONAL STATUS FIELD - define observed state of cluster
	// Important: Run "make" to regenerate code after modifying this file
}

//+kubebuilder:object:root=true
//+kubebuilder:subresource:status

// PostgreSQLServiceUser is the Schema for the postgresqlserviceusers API
type PostgreSQLServiceUser struct {
	metav1.TypeMeta   `json:",inline"`
	metav1.ObjectMeta `json:"metadata,omitempty"`

	Spec   PostgreSQLServiceUserSpec   `json:"spec,omitempty"`
	Status PostgreSQLServiceUserStatus `json:"status,omitempty"`
}
```

File: api/v1alpha1/postgresqlserviceuser_types.go (hunk 4 above)
Comment (kjuulh) [reply api/v1alpha1/postgresqlserviceuser_types.go]: it should be implemented now

File: api/v1alpha1/postgresqlserviceuser_types.go [hunk 6]
```diff
@@ -17,17 +17,15 @@ limitations under the License.
...
 
-// EDIT THIS FILE!  THIS IS SCAFFOLDING FOR YOU TO OWN!
-// NOTE: json tags are required.  Any new fields you add must have json tags for the fields to be serialized.
-
 // PostgreSQLServiceUserSpec defines the desired state of PostgreSQLServiceUser
 // +k8s:openapi-gen=true
 type PostgreSQLServiceUserSpec struct {
 	// Name of the service user
-	Name string `json:"name"`
+	Username string `json:"username"`
```
Comment (JacobValdemar) [inline api/v1alpha1/postgresqlserviceuser_types.go:28]: Consider if `username` also should be of type `ResourceVar` so all of it can come from a secret.

**Review by JacobValdemar — APPROVED** [review]
Nice improvements  🚀

**Pull request merged**